### ISSUE TITLE
Encoder-Decoder Based Seq-to-Seq Modelling

### ISSUE THREAD

**Issue by talhahashim** [talhahashim/Encoder-Decoder-Based-Seq-to-Seq-Modelling-#1]
# c_syntax_seq2seq.py

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader

# --------- CONFIG ---------
EMBEDDING_DIM = 128
HIDDEN_DIM = 256
TEACHER_FORCING_RATIO = 0.5
MAX_LEN = 30
PAD_IDX = 0
BATCH_SIZE = 32
EPOCHS = 20
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# --------- TOY VOCAB ---------
vocab = {f'tok{i}': i for i in range(130)}  # Simulated vocab of size 130
vocab['<PAD>'] = PAD_IDX
vocab['<SOS>'] = 1
vocab['<EOS>'] = 2
vocab['<UNK>'] = 3
vocab_size = 130
inv_vocab = {v: k for k, v in vocab.items()}

# --------- ENCODER ---------
class Encoder(nn.Module):
    def __init__(self, vocab_size, embed_dim, hidden_dim):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, embed_dim, padding_idx=PAD_IDX)
        self.lstm = nn.LSTM(embed_dim, hidden_dim, batch_first=True)

    def forward(self, src):
        embedded = self.embedding(src)
        outputs, (hidden, cell) = self.lstm(embedded)
        return hidden, cell

# --------- DECODER ---------
class Decoder(nn.Module):
    def __init__(self, vocab_size, embed_dim, hidden_dim):
        super().__init__()
        self.embedding = nn.Embedding(vocab_size, embed_dim, padding_idx=PAD_IDX)
        self.lstm = nn.LSTM(embed_dim, hidden_dim, batch_first=True)
        self.fc_out = nn.Linear(hidden_dim, vocab_size)

    def forward(self, input_token, hidden, cell):
        embedded = self.embedding(input_token.unsqueeze(1))  # [B, 1, E]
        output, (hidden, cell) = self.lstm(embedded, (hidden, cell))
        pred = self.fc_out(output.squeeze(1))  # [B, vocab_size]
        return pred, hidden, cell

# --------- SEQ2SEQ ---------
class Seq2Seq(nn.Module):
    def __init__(self, encoder, decoder):
        super().__init__()
        self.encoder = encoder
        self.decoder = decoder

    def forward(self, src, trg, teacher_forcing_ratio=0.5):
        B, T = trg.shape
        outputs = torch.zeros(B, T, vocab_size).to(DEVICE)

        hidden, cell = self.encoder(src)
        input_token = trg[:, 0]  # <SOS>

        for t in range(1, T):
            output, hidden, cell = self.decoder(input_token, hidden, cell)
            outputs[:, t] = output
            teacher_force = torch.rand(1).item() < teacher_forcing_ratio
            top1 = output.argmax(1)
            input_token = trg[:, t] if teacher_force else top1

        return outputs

# --------- DUMMY DATASET ---------
class CSyntaxDataset(Dataset):
    def __init__(self, samples):
        self.samples = samples  # list of (incorrect_seq, correct_seq)

    def __len__(self):
        return len(self.samples)

    def __getitem__(self, idx):
        inc, cor = self.samples[idx]
        inc = self.pad_sequence(inc)
        cor = self.pad_sequence([vocab['<SOS>']] + cor + [vocab['<EOS>']])
        return torch.tensor(inc), torch.tensor(cor)

    def pad_sequence(self, seq):
        return seq + [PAD_IDX] * (MAX_LEN - len(seq))

# --------- TRAINING LOOP ---------
def train(model, dataloader, optimizer, criterion):
    model.train()
    total_loss = 0
    for src, trg in dataloader:
        src, trg = src.to(DEVICE), trg.to(DEVICE)
        optimizer.zero_grad()
        output = model(src, trg, TEACHER_FORCING_RATIO)
        output_dim = output.shape[-1]
        output = output[:, 1:].reshape(-1, output_dim)
        trg = trg[:, 1:].reshape(-1)
        loss = criterion(output, trg)
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    return total_loss / len(dataloader)

# --------- INFERENCE ---------
def greedy_decode(model, src):
    model.eval()
    with torch.no_grad():
        src = torch.tensor(src).unsqueeze(0).to(DEVICE)
        hidden, cell = model.encoder(src)
        input_token = torch.tensor([vocab['<SOS>']]).to(DEVICE)
        outputs = []
        for _ in range(MAX_LEN):
            output, hidden, cell = model.decoder(input_token, hidden, cell)
            top1 = output.argmax(1)
            if top1.item() == vocab['<EOS>']:
                break
            outputs.append(top1.item())
            input_token = top1
        return [inv_vocab.get(tok, '<UNK>') for tok in outputs]

# --------- MAIN ---------
if __name__ == '__main__':
    # Sample dummy data
    samples = [
        ([4, 5, 12, 6, 7], [4, 5, 12, 6, 7, 8]),  # if(x){ --> if(x){}
        ([12, 10, 13, 9], [12, 10, 13, 9])         # x=y; --> x=y;
    ]

    dataset = CSyntaxDataset(samples)
    dataloader = DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=True)

    encoder = Encoder(vocab_size, EMBEDDING_DIM, HIDDEN_DIM).to(DEVICE)
    decoder = Decoder(vocab_size, EMBEDDING_DIM, HIDDEN_DIM).to(DEVICE)
    model = Seq2Seq(encoder, decoder).to(DEVICE)

    optimizer = optim.Adam(model.parameters(), lr=1e-3)
    criterion = nn.CrossEntropyLoss(ignore_index=PAD_IDX)

    for epoch in range(EPOCHS):
        loss = train(model, dataloader, optimizer, criterion)
        print(f"Epoch {epoch+1}: Loss = {loss:.4f}")

    # Inference
    test_seq = [4, 5, 12, 6, 7]  # if(x){
    test_seq = test_seq + [PAD_IDX] * (MAX_LEN - len(test_seq))
    corrected = greedy_decode(model, test_seq)
    print("Predicted Fix:", ' '.join(corrected))
